toolchain/sem_ir/name_scope.h

```diff
@@ -18,13 +18,104 @@ enum class AccessKind : int8_t {
+  // Depending on the given `inst_id`, it can be:
+  // * Found: Lookup was successful and `inst_id` is valid. If `inst_id` is
+  // known to be valid, use `CreateFound()`.
+  // * Not Found: Lookup returned no results and `inst_id` is invalid. If
+  // `inst_id` is known to be invalid, use `CreateNotFound()`.
+  // * Error: Lookup encountered an error and `inst_id` is
+  // `ErrorInst::SingletonInstId`, which is considered valid and found. If
+  // `inst_id` is known to be `ErrorInst::SingletonInstId`, `use CreateError()`.
+  //
+  // To create a poisoned result, use `CreatePoisoned()`.
+  static auto Create(InstId inst_id, AccessKind access_kind)
+      -> ScopeLookupResult {
+    return ScopeLookupResult(inst_id, access_kind, false);
+  }
+
+  // Creates a successful lookup result. `inst_id` must be valid. If
+  // `inst_id` is known to be `ErrorInst::SingletonInstId`, `use CreateError()`.
+  static auto CreateFound(InstId inst_id, AccessKind access_kind)
+      -> ScopeLookupResult {
+    CARBON_CHECK(inst_id.is_valid());
+    return Create(inst_id, access_kind);
+  }
+
+  // Creates a result when there are no result.
+  static auto CreateNotFound() -> ScopeLookupResult {
+    return Create(InstId::Invalid, AccessKind::Public);
+  }
+
+  // Creates a poisoned result.
+  static auto CreatePoisoned() -> ScopeLookupResult {
+    return ScopeLookupResult(InstId::Invalid, AccessKind::Public, true);
+  }
+
+  // Creates a result when an error was encountered, which is considered valid
+  // and found.
+  static auto CreateError() -> ScopeLookupResult {
+    return CreateFound(ErrorInst::SingletonInstId, AccessKind::Public);
+  }
+
+  // True iff CreatePoisoned() was used.
+  auto is_poisoned() const -> bool { return is_poisoned_; }
+
+  // True when lookup was successful or resulted with an error.
+  // False for poisoned, not found.
+  auto is_found() const -> bool {
+    return !is_poisoned() && lookup_inst_id().is_valid();
+  }
+
+  // The `InstId` of the result of the lookup.
+  // Must not be called for poisoned results.
+  // If valid, it means the lookup was successful or resulted with an error.
+  auto lookup_inst_id() const -> InstId {
+    CARBON_CHECK(!is_poisoned());
+    return inst_id_;
+  }
+
+  // The `InstId` of the result of the lookup.
+  // Must only be called when lookup was successful.
+  auto found_inst_id() const -> InstId {
+    CARBON_CHECK(is_found());
+    return inst_id_;
+  }
+
+  auto access_kind() const -> AccessKind { return access_kind_; }
+
+  friend auto operator==(const ScopeLookupResult&, const ScopeLookupResult&)
+      -> bool = default;
+
+ private:
+  explicit ScopeLookupResult(InstId inst_id, AccessKind access_kind,
+                             bool is_poisoned)
+      : inst_id_(inst_id),
+        access_kind_(access_kind),
+        is_poisoned_(is_poisoned) {}
+
+  InstId inst_id_ = InstId::Invalid;
+  AccessKind access_kind_ = AccessKind::Public;
+  bool is_poisoned_ = false;
```

Comment:
```suggestion
  InstId inst_id_;
  AccessKind access_kind_;
  bool is_poisoned_;
```

The only constructor specifies these, so I don't think these are used -- I think it's best not to provide unused defaults.

Reply:
Done, but I wholeheartedly disagree :)
Code changes and we could forget to initialize a data member explicitly in the constructor, or wrongly assume it's initialized to a specific value using a default constructor, so always having a default value prevents undefined behavior.

Reply:
It's interesting you point to the constructor as a reason... to give you a different perspective, while I can audit a constructor and make sure members are initialized, the compiler can also validate that members are assigned -- unless a default is added. So to me, adding a default value that's not intended to be used is actually _removing_ both human and compiler safety checks.

Reply:
Sure, if we know that the compiler will break compilation if a data member is not initialized then we definitely don't need this and I see your point about make this less readable.
I remember at least one case of a serious production breakage where adding a default value for a data member fixed it. Compilers have advanced since then so it's possible this is no longer necessary.